httpGw: make /tcp/0 work

cryptix opened the pull request:
Prints out the chosen port and puts it into the config so the cli works over it too.

Fixes #1288

cryptix commented on core/corehttp/corehttp.go:
```diff
@@ -67,6 +68,17 @@ func listenAndServe(node *core.IpfsNode, addr ma.Multiaddr, handler http.Handler
 		return err
 	}
 
+	host, port, err := net.SplitHostPort(list.Addr().String())
```
is there a helper somewhere to construct a multiaddr form a `net.Addr` directly?

cryptix commented:
Still WIP, After talking with @krl and @jbenet, we want to see if we can put the address in the `repo.lock`.

Would require centralizing the `net.Listen()` calls in one place and adding a new corehttp config option.

jbenet commented on cmd/ipfs/daemon.go:
```diff
@@ -258,28 +254,59 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
+	if err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
+	}
```
sure it's ok to move this check down? not sure if it was up above for a reason (i.e. bail out early before something else)

cryptix replied:
Hrm.. that's a good point - my idea was to group the bring-up of fuse mount, gateway and API. I can revert it if you prefer the early-error out. I'm not sure if daemon startup failure something you want to optimizer over clarity for though.

jbenet replied:
agree with you. if there's no error here, let's move it down.

jbenet commented on cmd/ipfs/daemon.go:
```diff
@@ -258,28 +254,59 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
+	if err := node.Repo.SetConfigKey("Addresses.API", apiMaddr.String()); err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
+	}
```
not super comfortable merging with this. why don't we make the change of writing the api address to a special file in the repo, _and then_ implement `/tcp/0` support

cryptix replied:
Would you be fine without persisting to config and doing that in a follow up?
I guess it's to mingled but the main idea here was to print only after the listener opened correctly.

jbenet replied:
> I guess it's to mingled but the main idea here was to print only after the listener opened correctly.

Ahhh. yes, that is a good thing. could we have only that in this PR? how about:

print after open --> put addr in file in repo --> support `/tcp/0` ?

jbenet replied:
or, ok i guess it's fine to support `/tcp/0` if we don't persist to config.

cryptix commented:
rebased

cryptix commented:
rebased (2)

jbenet commented:
Running tests on all commits

cryptix commented:
squashed and rebased onto master.

i reorderd it to do api first, gateway 2nd and had to change the way `corehttp.Serve` errors are gathered. I like this more though, gateway http errors were only logged and not returend.

jbenet commented:
@cryptix would you mind splitting this into function for the steps? one big function is scary to edit because not clear what is required to be in that order. functions are nicer because smaller units of work are easier to reason about. you have this paged into your mind right now and are best suited for it.

cryptix commented:
rebased for fun and test results

cryptix commented:
@jbenet i reworked the helper to not goroutine until necessary, this way they don't race on writing to stdout (this caused the weird 'daemon output not okay' test faults i belive)

RFM if its fine with you.

whyrusleeping commented on cmd/ipfs/daemon.go:
```diff
@@ -192,98 +195,76 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
-	if len(cfg.Gateway.RootRedirect) > 0 {
-		rootRedirect = corehttp.RedirectOption("", cfg.Gateway.RootRedirect)
+// mountHTTPapi collects options, creates listener, prints status message and starts serving requests
+func mountHTTPapi(req cmds.Request) (error, <-chan error) {
```
calling it `mount` might confuse things with fuse

cryptix replied:
hm.. yea.. `open` or `listenAndServe`?

jbenet replied:
maybe `serveHTTPApi`

jbenet commented on cmd/ipfs/daemon.go:
```diff
@@ -300,18 +281,146 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
+}
+
+// mountHTTPgw collects options, creates listener, prints status message and starts serving requests
+func mountHTTPgw(req cmds.Request) (error, <-chan error) {
```
`serveHTTPGateway`

jbenet replied:
(should it be `serveHttpGateway`? dont recall what's go idiomatic)

jbenet commented on cmd/ipfs/daemon.go:
```diff
@@ -300,18 +281,146 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		if c != nil {
```
this is incorrect. if any item in `cs` is nil, this will block forever because of `Add(len(cs))`. 

let's stick to the:

```go
{
  wg.Add(1)
  go output(c)
}
```
idiom.

jbenet commented:
@cryptix more comments o/

cryptix commented:
rebased and addressed comments

jbenet commented:
LGTM i'll merge after tests run
